pkg/fdsrv/fdsrv.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2022 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Serves a file descriptor over an AF_UNIX socket when presented with a nonce.
+//
+// You must pass the socket path and nonce to the client via some out-of-band
+// mechanism, such as gRPC or a bash script.
+//
+// Notes:
+// - Uses the unix domain socket abstract namespace
+// - Picks its own path in the abstract namespace for the socket.
+// - Shared FDs are essentially duped, and they point to the same struct file:
+// they share offsets and whatnot.
+//
+// Options:
+// - WithServeOnce: serve once and shuts down (default is forever)
+// - WithTimeout: cancel itself after a timeout (default none)
+//
+// Usage Server:
+//
+//	fds, err := NewFdServer(fd_to_share, "some_nonce", WithServeOnce())
+//	var s path = fds.UdsPath()
+//
+//	// Pass path and some_nonce to the client via an out of band mechanism
+//
+//	fds.Serve(); // Blocks until the server is done
+//	fds.Destroy()
+//
+// Usage Client:
+//
+//	sfd, err := GetSharedFd("uds_path", "some_nonce")
+package fdsrv
+
+import (
+	"errors"
+	"io"
+	"net"
+	"syscall"
+	"time"
+)
+
+type FdServer struct {
+	dupedFd   int
+	nonce     string
+	listener  *net.UnixListener
+	timeout   time.Duration
+	serveOnce bool
+}
+
+// Serves the fd, returns true if successful, err for a server error.
+// "false, nil" means the client was wrong, not the server.
+func (fds *FdServer) handleConnection(uc *net.UnixConn) (bool, error) {
+	defer uc.Close()
+
+	buf := make([]byte, 4096)
+	n, err := uc.Read(buf)
+	if err != nil {
+		return false, err
+	}
+	query := string(buf[:n])
+	if query != fds.nonce {
+		io.WriteString(uc, "BAD NONCE")
+		return false, nil
+	}
+	oob := syscall.UnixRights(fds.dupedFd)
+	good := []byte("GOOD NONCE")
+	goodn, oobn, err := uc.WriteMsgUnix(good, oob, nil)
+	if err != nil {
+		return false, err
+	}
+	if goodn != len(good) || oobn != len(oob) {
+		return false, errors.New("truncated write")
+	}
+	return true, nil
+}
+
+func NewFdServer(fd int, nonce string,
+	options ...func(*FdServer) error) (*FdServer, error) {
+	var err error
+	fds := &FdServer{}
+
+	if len(nonce) == 0 {
+		return nil, errors.New("nonce must not be empty")
+	}
+	fds.nonce = nonce
+
+	for _, op := range options {
+		if err := op(fds); err != nil {
+			return nil, err
+		}
+	}
+
+	// An empty addr tells Linux to "autobind" to an available path in the
+	// abstract unix domain socket namespace
+	ua, err := net.ResolveUnixAddr("unix", "")
+	if err != nil {
+		return nil, err
+	}
+	fds.listener, err = net.ListenUnix("unix", ua)
+	if err != nil {
+		return nil, err
+	}
+
+	// Caller could close the file while we are running.  Keep our own copy.
+	fds.dupedFd, err = syscall.Dup(int(fd))
+	if err != nil {
+		fds.listener.Close()
+		return nil, err
+	}
+
+	return fds, nil
+}
+
+func WithTimeout(timeout time.Duration) func(*FdServer) error {
```

Comment:
The other convention would be, for all exported stuff, to have a comment:
// WIthTimeOut adds a timeout option to a Server.